Use context appropriately

abicky opened the pull request:
This is the subsequent PR of https://github.com/datachainlab/ethereum-ibc-relay-prover/pull/32 and replaces `context.TODO()` with an appropriate context. 

I have confirmed that the CI of https://github.com/datachainlab/cosmos-ethereum-ibc-lcp passes:

https://github.com/datachainlab/cosmos-ethereum-ibc-lcp/actions/runs/13850650603/job/38757230693

abicky committed:
Update ethereum-ibc-relay-chain

Signed-off-by: abicky <takeshi.arabiki@datachain.jp>

abicky committed:
Make beacon package support context

Signed-off-by: abicky <takeshi.arabiki@datachain.jp>

abicky committed:
Make relay package use context appropriately

Signed-off-by: abicky <takeshi.arabiki@datachain.jp>

abicky commented on relay/execution.go:
```diff
@@ -1,15 +1,17 @@
 
-func (pr *Prover) buildAccountUpdate(blockNumber uint64) (*lctypes.AccountUpdate, error) {
+func (pr *Prover) buildAccountUpdate(ctx context.Context, blockNumber uint64) (*lctypes.AccountUpdate, error) {
 	proof, err := pr.executionClient.GetProof(
```
The signature has changed in the PR https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/64#discussion_r1978699763.

siburu commented on beacon/client.go:
```diff
@@ -101,9 +103,14 @@ func (cl Client) GetLightClientFinalityUpdate() (*LightClientFinalityUpdateRespo
+		return err
+	}
+
+	r, err := cl.httpClient.Do(req)
```
@abicky `http.Get` is equivalent to `http.DefaultClient.Get`.
ref: https://cs.opensource.google/go/go/+/refs/tags/go1.24.1:src/net/http/client.go;l=452-454

Therefore it is more suitable to use `http.DefaultClient.Do` here, right?

abicky replied:
I'm afraid `http.Get` doesn't support a context because we need to use a `http.NewRequestWithContext` to pass a context.

abicky replied:
Ah, I understood your point. Umm, I will think of using `http.DefaultClient` but currently I feel it is better to create a dedicate HTTP client.

abicky replied:
@siburu To conclude, in my opinion, we should use a dedicated client and set `Timeout` to, for example, 5 minutes.
What do you think about this?

I will explain the reason below.

In general, it is recommended to avoid using `http.DefaultClient` as follows:

* To set `Timeout`
    - The `Timeout` of the default client is 0, which means there is no timeout
* To increase `MaxIdleConnsPerHost`
    - The default value is 2 and it is not sufficient a process send many concurrent requests
* To avoid being affected by idle connections
    - The default value is 100, and if a process sends requests to various services, a large portion of the idle connections might be allocated to a particular service

Here are some blog posts that suggest we avoid using `http.DefaultClient`:

- [Don’t use Go’s default HTTP client (in production)](https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779)
- [Know when to break up with Go's http.DefaultClient](https://vishnubharathi.codes/blog/know-when-to-break-up-with-go-http-defaultclient/)
- [[Go] 前方互換性を保ちながらhttp.DefaultTransportからチューニングしたhttp.Transportをつくる](https://budougumi0617.github.io/2021/09/13/how_to_copy_default_transport/) (in Japanese)

As for `beacon.Client`, we probably won't adjust parameters except for `Timeout` in the near future because the concurrency is currently quite low, but we should set `Timeout` as follows:

```diff
diff --git a/beacon/client.go b/beacon/client.go
index 307eb3d..55077d0 100644
--- a/beacon/client.go
+++ b/beacon/client.go
@@ -8,6 +8,7 @@ import (
        "io"
        "net/http"
        "slices"
+       "time"

        "github.com/hyperledger-labs/yui-relayer/log"
 )
@@ -20,7 +21,12 @@ type Client struct {
 }

 func NewClient(endpoint string) Client {
-       return Client{endpoint: endpoint, httpClient: &http.Client{}}
+       return Client{
+               endpoint: endpoint,
+               httpClient: &http.Client{
+                       Timeout: 5 * time.Minute,
+               },
+       }
 }

 func IsSupportedVersion(v string) bool {
```

bluele replied:
For timeout, in recent Go(since 1.7?), I think it is more common to use `context.WithTimeout` rather than setting `Client.Timeout` at the client level. 
The context-based approach allows for request-specific timeout configurations, providing more flexibility and controllability over request handling.

siburu replied:
@bluele 
> The context-based approach allows for request-specific timeout configurations

Sure. Thank you.

@abicky 
If we need to use a non-default client (for example, a client with non-default idle connections) in this PR, it's OK.
Otherwise, we should continue to use `http.DefaultClient` for now, right?

abicky replied:
@bluele @siburu Thank you for your feedback! It might be better to have a default timeout, but I reconsidered that if we introduce it we should also make the timeout configurable. So, if we introduce such a feature, we should work on another pull request.

I have fixed the code in https://github.com/datachainlab/ethereum-ibc-relay-prover/pull/35/commits/82c04273608cfe8a242d37e05574215e88701997.

abicky committed:
Use http.DefaultClient

cf. https://github.com/datachainlab/ethereum-ibc-relay-prover/pull/35/files#r1998753906

Signed-off-by: Takeshi Arabiki <takeshi.arabiki@datachain.jp>

siburu reviewed (approved):
@abicky LGTM, thank you!!!
